federation/pkg/federation-controller/sync/controller.go

```diff
@@ -215,17 +221,55 @@ func (s *FederationSyncController) Run(stopChan <-chan struct{}) {
 	go func() {
 		<-stopChan
 		s.informer.Stop()
+		s.workQueue.ShutDown()
 	}()
 	s.deliverer.StartWithHandler(func(item *util.DelayingDelivererItem) {
-		namespacedName := item.Value.(*types.NamespacedName)
-		s.reconcile(*namespacedName)
+		s.workQueue.Add(item)
 	})
 	s.clusterDeliverer.StartWithHandler(func(_ *util.DelayingDelivererItem) {
 		s.reconcileOnClusterChange()
 	})
+
+	// TODO: Allow multiple workers.
+	go wait.Until(s.worker, time.Second, stopChan)
+
 	util.StartBackoffGC(s.backoff, stopChan)
 }
 
+type reconciliationStatus int
+
+const (
+	statusAllOK reconciliationStatus = iota
+	statusNeedsRecheck
+	statusError
+	statusNotSynced
+)
+
+func (s *FederationSyncController) worker() {
+	for {
+		obj, quit := s.workQueue.Get()
+		if quit {
+			return
+		}
+
+		item := obj.(*util.DelayingDelivererItem)
+		namespacedName := item.Value.(*types.NamespacedName)
+		status := s.reconcile(*namespacedName)
+		s.workQueue.Done(item)
+
+		switch status {
+		case statusAllOK:
+			break
+		case statusError:
+			s.deliver(*namespacedName, 0, true)
+		case statusNeedsRecheck:
+			s.deliver(*namespacedName, s.reviewDelay, false)
+		case statusNotSynced:
+			s.deliver(*namespacedName, s.reviewDelay, false)
```

Comment:
I didn't notice until I went to rebase on this, but shouldn't this be s.ClusterAvailableDelay?